---
title: "AD Ortholog Exploration"
author: "Alexander Bergendorf"
date: "2023-08-07"
output: html_document
---


# Library Import
```{r}
library(Seurat)
library(mixOmics)
library(tidyverse)
library(orthogene)
library(dittoSeq)
library(openxlsx)
#library(PCAtools)
library(MASS)
library(glmnet)
library(pROC)
#home_dir <- "C:/Users/aberg/OneDrive/Desktop/Research_Brubaker/AD TransComp-R Project Summer-Fall 2023/Brubaker Summer Research 2023"
home_dir <- "C:/Users/Alexander/OneDrive/Desktop/Research_Brubaker/AD TransComp-R Project Summer-Fall 2023/Brubaker Summer Research 2023"
```

# PCA Exploration
## Loading Mouse Data
```{r}
combined <- readRDS(file = paste0(home_dir, "/AD_mouse_objects_ortholog/mouse_adwt_labeled.RData"))
combined <- subset(combined, subset = summary == "Microglia")
```




## Run PCA to isolate mouse Microglia
```{r}
DefaultAssay(combined) <- "RNA"
combined <- NormalizeData(combined)
combined <- FindVariableFeatures(combined, selection.method = "vst", nfeatures = 3000)
all.genes <- rownames(combined)
combined <- ScaleData(combined, features = all.genes)
combined <- RunPCA(combined, features = VariableFeatures(object = combined))
```


```{r}
dittoDimPlot(combined, var = "Genotype", reduction.use = "pca")
```

```{r}
DimHeatmap(combined, dims = 1:4)
```

A lot of these principal components are separated by only a few cells... I think we need to cluster the cells again to get a better representation 

```{r}
combined <- FindNeighbors(combined, graph.name = "test", dims = 1:5)
combined <- FindClusters(combined, graph.name = "test", resolution = 0.5)
combined <- RunUMAP(combined, graph.name = "test", dims = 1:5)
```


```{r}
dittoDimPlot(combined, var = "Genotype", reduction = "umap")
dittoDimPlot(combined, "seurat_clusters", do.label = TRUE, legend.show = FALSE, labels.size = 3) 
dittoDimPlot(combined, var = "CellType", reduction = "umap")
```


```{r}
#markers <- FindAllMarkers(combined, assay = "RNA", only.pos = TRUE, logfc.threshold = 0.5, test.use = "MAST", latent.vars = "orig.ident") 

#clust_counts <- table(combined@meta.data[,"seurat_clusters"]) 

#clust_markers_list <- list() 

#for(i in unique(markers$cluster)){ 

#clustname <- paste0('cluster', i, '_N', clust_counts[i]) 

#clust_markers_list[[clustname]] <- subset(markers, markers$cluster == i) 

#rownames(clust_markers_list[[clustname]]) <- clust_markers_list[[clustname]]$gene 

#} 

#write.xlsx(clust_markers_list, file = paste0(home_dir, "/mouse_microglia_reclustering.xlsx"))
```



```{r}
combined <- FindSubCluster(
  combined,
  5,
  "test",
  subcluster.name = "sub.cluster",
  resolution = 0.4
)
```

```{r}
dittoDimPlot(combined, var = "sub.cluster", reduction = "umap")
dittoDimPlot(combined, var = "CellType", reduction = "umap")
```


```{r}
# Cluster 5 and 6 are all non-microglial cels
combined <- subset(combined, subset = sub.cluster %in% c("5_1", "5_2", "6", "7"), invert = TRUE)
```



```{r}
# Save barcodes of good microglial cells for future use
good_microglia <- Cells(combined)
#saveRDS(good_microglia, file = paste0(home_dir, "/AD_mouse_objects_ortholog/good_microglia_barcodes.RData"))
```


## Rerun PCA mouse
```{r}
# Rerunning PCA after filtering
DefaultAssay(combined) <- "RNA"
combined <- NormalizeData(combined)
combined <- FindVariableFeatures(combined, selection.method = "vst", nfeatures = 3250)
all.genes <- rownames(combined)
combined <- ScaleData(combined, features = all.genes)
combined <- RunPCA(combined, features = VariableFeatures(object = combined))
```


```{r}
# Variances of mosue PCs
stdev <- combined@reductions$pca@stdev
variance <- stdev^2
variance <- variance / sum(variance)
variance
#write.csv(variance, paste0(home_dir, "/pca_variances_mouse.csv"))
```

```{r}
variance <- read.csv(paste0(home_dir, "/pca_variances_mouse.csv"))
colnames(variance) <- c("PC", "sdev")
variance$cum_var <- cumsum(variance$sdev)
variance
```

```{r}
svg('C:/Users/aberg/OneDrive/Desktop/Mouse_variance_explained_pca.svg', width = 5, height = 4)
ggplot(data = variance, aes(x = PC, y = cum_var, group = 1)) +
  geom_line(color = "black") +  # Changed line color to blue
  geom_point(color = "black") +  # Added points with the same color
  geom_hline(yintercept = 0.80, linetype = "dashed", color = "red") +  # Added red dashed line at 0.80 cum_var
  geom_point(data = subset(variance, cum_var == 0.80), aes(x = PC, y = cum_var), color = "red", size = 3) +  # Add red point at the threshold
  ylab("Cumulative Variance") +# Changed y-axis label
  theme_classic() + theme_bw()
dev.off()
```


## Figure 2B
```{r}
# Cluster cells and visualize on UMAP for paper visualization
combined <- FindNeighbors(combined, dims = 1:5)
combined <- FindClusters(combined, resolution = 0.5)
combined <- RunUMAP(combined, dims = 1:5)
```


```{r}
p1 <- dittoDimPlot(combined, "Genotype", color.panel = c("#F8766D", "#00BFC4"))
cairo_ps(filename= paste0(home_dir, "/Manuscript and Figures/Figure 2/2B.eps"), width = 9, height = 7) # name the file as an EPS
p1 #variable of the figure that is defined/saved, can be used any name
dev.off()
```


```{r}
# Counts of mouse microglial cells
count_table <- table(combined@meta.data$Genotype)
count_table
#write.csv(count_table, file = paste0(home_dir, "/Manuscript and Figures/Figure 2/mouse_counts.csv"))
```
```{r}
table(combined@meta.data$SampleID)
```


```{r}
ElbowPlot(combined, ndims = 50)
```


```{r}
# Pull principal components from Seurat object
principal_components <- combined@reductions$pca@feature.loadings
head(principal_components)
principal_components <- as.data.frame(principal_components)
```

Now that we have the eigenvectors/eigenvalues of the PCA covariance matrix, lets bring in the human data and project the data into mouse PC space

## Run PCA to isolate human microglia
```{r}
human <- readRDS(file = paste0(home_dir, "/AD_human_objects_ortholog/human_adwt_labeled_ortholog.RData"))
human <- subset(human, subset = summary == "Microglia")
```


```{r}
# Run initial PCA
DefaultAssay(human) <- "RNA"
human <- NormalizeData(human)
human <- FindVariableFeatures(human, selection.method = "vst", nfeatures = 3000)
all.genes <- rownames(human)
human <- ScaleData(human, features = all.genes)
human <- RunPCA(human, features = VariableFeatures(object = human))
```

```{r}
ElbowPlot(human)
```

```{r}
# Same case as for Mouse, few cells causing separation, so, let's filter non-microglial cells out
dittoDimPlot(human, var = "Genotype", reduction.use = "pca")
```

```{r}
# Clustering 
human <- FindNeighbors(human, dims = 1:10)
human <- FindClusters(human, resolution = 0.5)
human <- RunUMAP(human, dims = 1:10)
```

```{r}
dittoDimPlot(human, var = "Genotype", reduction = "umap")
dittoDimPlot(human, "seurat_clusters", do.label = TRUE, legend.show = FALSE, labels.size = 3) 
```

```{r}
#markers <- FindAllMarkers(human, assay = "RNA", only.pos = TRUE, logfc.threshold = 0.5, test.use = "MAST", latent.vars = "orig.ident") 

#clust_counts <- table(human@meta.data[,"seurat_clusters"]) 

#clust_markers_list <- list() 

#for(i in unique(markers$cluster)){ 

#clustname <- paste0('cluster', i, '_N', clust_counts[i]) 

#clust_markers_list[[clustname]] <- subset(markers, markers$cluster == i) 

#rownames(clust_markers_list[[clustname]]) <- clust_markers_list[[clustname]]$gene 

#} 

#write.xlsx(clust_markers_list, file = paste0(home_dir, "/human_microglia_reclustering.xlsx"))
```


```{r}
# Based on the markers, 6, 7, 9, 10, and 11 are not microglial cells
human <- subset(human, subset = seurat_clusters %in% c(6, 7, 9, 10, 11), invert = TRUE)
```

```{r}
dittoDimPlot(human, var = "Genotype", reduction = "umap")
dittoDimPlot(human, "seurat_clusters", do.label = TRUE, legend.show = FALSE, labels.size = 3) 
```


```{r}
# Saving barcodes of good human microglia
good_microglia <- Cells(human)
#saveRDS(good_microglia, file = paste0(home_dir, "/AD_human_objects_ortholog/good_microglia_barcodes.RData"))
```



## Get Human GEX matrix
```{r}
# Rerunning PCA again
DefaultAssay(human) <- "RNA"
human <- NormalizeData(human)
all.genes <- rownames(human)
human <- ScaleData(human, features = all.genes)
human_gex_matrix <- as.data.frame(human@assays$RNA@scale.data)
human <- RunPCA(human, features = VariableFeatures(object = human))
```

## Figure 2D
```{r}
# Plotting on UMAP for visualization
human <- FindNeighbors(human, dims = 1:5)
human <- FindClusters(human, resolution = 0.5)
human <- RunUMAP(human, dims = 1:5)
```

```{r}
p2 <- dittoDimPlot(human, "Genotype", color.panel = c("#F8766D", "#00BFC4"))
cairo_ps(filename= paste0(home_dir, "/Manuscript and Figures/Figure 2/2D.eps"), width = 9, height = 7) # name the file as an EPS
p2 #variable of the figure that is defined/saved, can be used any name
dev.off()
```

```{r}
# Count of human microglial cells
count_table <- table(human@meta.data$Genotype)
count_table
#write.csv(count_table, file = paste0(home_dir, "/Manuscript and Figures/Figure 2/human_counts.csv"))
```

```{r}
table(human@meta.data$Sample)
```



```{r}
# Confirming only orthologs in GEX matrix
ortholog_mat <- convert_orthologs(gene_df = row.names(human_gex_matrix),
                                    input_species = "human",
                                    output_species = "mouse",
                                    non121_strategy = "drop_both_species")

# Filtering the orthologs based on the 3250 principal component loadings
ortholog_mat <- ortholog_mat %>%
  filter(row.names(ortholog_mat) %in% row.names(principal_components))
        

# Filtering the human gex matrix based on the 3250 principal component loadings
human_gex_matrix <- human_gex_matrix %>%
  filter(row.names(human_gex_matrix) %in% ortholog_mat$input_gene)


# Ordering human_gex_matrix in the same manner as the pc loadings
row.names(human_gex_matrix) <- row.names(ortholog_mat)

# Arrange alphabetically 
principal_components  <- principal_components %>% arrange(row.names(principal_components))

# Arrange alphabetically
human_gex_matrix <- human_gex_matrix %>%
  arrange(row.names(human_gex_matrix))

```


```{r}
# Save mouse object with pca loadings
#saveRDS(combined, file = paste0(home_dir, "/AD_mouse_objects_ortholog/mouse_pca_loadings.RData"))
```








```{r}
# converting to matrix
principal_components <- as.matrix(principal_components)
human_gex_matrix <- as.matrix(human_gex_matrix)

# assigning human names to pc matrix
principal_components <- principal_components[row.names(human_gex_matrix),]

# Ensuring equal rows, equals 3000
ncol(t(principal_components))
nrow(human_gex_matrix)

# Multiplies the transpose of the principal component matrix by the human gex matrix
multiplication_2 <- t(human_gex_matrix) %*% principal_components
# Creates new dimensional reduction on human Seurat object with the pc projections
human[["pca_projections"]] <- CreateDimReducObject(embeddings = multiplication_2, key = "pca_proj")
```

```{r}
human@reductions$pca_projections
```



## GLMs
### Running GLM with PCs encpasulating 80% variance
```{r}
human <- readRDS(file = paste0(home_dir, "/AD_human_objects_ortholog/human_mouse_proj_pca.RData"))
```


```{r}
# Run GLM
ad.model <- glm(dv ~ pcaproj_1 + pcaproj_2 + pcaproj_3 + pcaproj_4 + pcaproj_5 + 
                  pcaproj_6 + pcaproj_7 + pcaproj_8 + pcaproj_9 + pcaproj_10 +
                  pcaproj_11 + pcaproj_12 + pcaproj_13 + pcaproj_14 + pcaproj_15 +
                  pcaproj_16 + pcaproj_17 + pcaproj_18 + pcaproj_19 + pcaproj_20 +
                  pcaproj_21 + pcaproj_22 + pcaproj_23 + pcaproj_24 + pcaproj_25 +
                  pcaproj_26 + pcaproj_27 + pcaproj_28 + pcaproj_29 + pcaproj_30 +
                  pcaproj_31 + pcaproj_32 + pcaproj_33 + pcaproj_34 + pcaproj_35 +
                  pcaproj_36
                  , data = human@meta.data, family = binomial)

```

```{r}
summary(ad.model)
```

```{r}
#saveRDS(ad.model, file = paste0(home_dir, "/AD_human_objects_ortholog/pca_glm_80var.RData"))
```


### AIC Optimization
```{r}
human <- readRDS(file = paste0(home_dir, "/AD_human_objects_ortholog/human_mouse_proj_pca.RData"))
```

```{r}
# Run GLM
ad.model <- glm(dv ~ pcaproj_1 + pcaproj_2 + pcaproj_3 + pcaproj_4 + pcaproj_5 + 
                  pcaproj_6 + pcaproj_7 + pcaproj_8 + pcaproj_9 + pcaproj_10 +
                  pcaproj_11 + pcaproj_12 + pcaproj_13 + pcaproj_14 + pcaproj_15 +
                  pcaproj_16 + pcaproj_17 + pcaproj_18 + pcaproj_19 + pcaproj_20 +
                  pcaproj_21 + pcaproj_22 + pcaproj_23 + pcaproj_24 + pcaproj_25 +
                  pcaproj_26 + pcaproj_27 + pcaproj_28 + pcaproj_29 + pcaproj_30 +
                  pcaproj_31 + pcaproj_32 + pcaproj_33 + pcaproj_34 + pcaproj_35 +
                  pcaproj_36
                  , data = human@meta.data, family = binomial)
```

```{r message=FALSE, include=FALSE}
final_model <- stepAIC(ad.model, direction = "both")
```

```{r}
summary(final_model)
```

```{r}
#saveRDS(final_model, file = paste0(home_dir, "/AD_human_objects_ortholog/pca_glm_aic_optimized.RData"))
final_model <- readRDS(file = paste0(home_dir, "/AD_human_objects_ortholog/pca_glm_aic_optimized.RData"))
summary(final_model)
```


### AIC Optimization w/ test/train 
```{r}
human_pca <- readRDS(file = paste0(home_dir, "/AD_human_objects_ortholog/human_mouse_proj_pca.RData"))
```


```{r}
set.seed(428)
all_data_pca <- human_pca@meta.data[,c(18:53, 68)]
sample_pca <- sample(c(TRUE,FALSE), nrow(all_data_pca),  
                 replace=TRUE, prob=c(0.8,0.2))
```



```{r}
# creating training dataset 
train_data_pca  <- all_data_pca[sample_pca, ] 
  
# creating testing dataset 
test_data_pca  <- all_data_pca[!sample_pca, ] 
```

```{r}
# Run GLM
ad.model <- glm(dv ~ pcaproj_1 + pcaproj_2 + pcaproj_3 + pcaproj_4 + pcaproj_5 + 
                  pcaproj_6 + pcaproj_7 + pcaproj_8 + pcaproj_9 + pcaproj_10 +
                  pcaproj_11 + pcaproj_12 + pcaproj_13 + pcaproj_14 + pcaproj_15 +
                  pcaproj_16 + pcaproj_17 + pcaproj_18 + pcaproj_19 + pcaproj_20 +
                  pcaproj_21 + pcaproj_22 + pcaproj_23 + pcaproj_24 + pcaproj_25 +
                  pcaproj_26 + pcaproj_27 + pcaproj_28 + pcaproj_29 + pcaproj_30 +
                  pcaproj_31 + pcaproj_32 + pcaproj_33 + pcaproj_34 + pcaproj_35 +
                  pcaproj_36
                  , data = train_data_pca, family = binomial)
```

```{r include=FALSE}
final_model <- stepAIC(ad.model, direction = "both")
```

```{r}
summary(final_model)
```


```{r}
predictions_pca <- predict(final_model, newdata = test_data_pca, type = "response")
```

```{r}
roc_pca <- roc(test_data_pca$dv, predictions_pca)
roc_pca
```

```{r}
plot(roc_pca, main = "ROC Curve (PCA Model)", col = "blue", lwd = 2, print.auc = TRUE)
```

### Lasso Selection
```{r}
human_pca <- readRDS(file = paste0(home_dir, "/AD_human_objects_ortholog/human_mouse_proj_pca.RData"))
```

```{r}
set.seed(428)
all_data_pca <- human_pca@meta.data[,c(18:53, 68)]
sample_pca <- sample(c(TRUE,FALSE), nrow(all_data_pca),  
                 replace=TRUE, prob=c(0.8,0.2))
```

```{r}
# creating training dataset 
train_data_pca  <- all_data_pca[sample_pca, ] 
  
# creating testing dataset 
test_data_pca  <- all_data_pca[!sample_pca, ] 
```

```{r}
predictors <- as.matrix(train_data_pca[,1:36])
response <- train_data_pca$dv
```


```{r}
# Determine lambda
cv.lasso <- cv.glmnet(predictors, response, family = "binomial")
plot(cv.lasso)
cv.lasso$lambda.min
coef(cv.lasso, s = "lambda.1se")
```


```{r}
# Final model with lambda.min
lasso.model <- glmnet(predictors, response, alpha = 1, family = "binomial",
                      lambda = cv.lasso$lambda.1se)
```


```{r}
predictions_pca <- predict.glmnet(lasso.model, newx = as.matrix(test_data_pca[,1:36]), type = "response")
```

```{r}
roc_pca <- roc(test_data_pca$dv, predictions_pca)
roc_pca
```



## Visualize PCs
```{r}
human <- readRDS(file = paste0(home_dir, "/AD_human_objects_ortholog/human_mouse_proj_spca.RData"))
```

```{r}
human@reductions$pca@cell.embeddings
human@reductions$spca_projections@cell.embeddings
```



### PC Plots
```{r}
# plot pc projections to visualize
for(i in 2:50){
p1 <- DimPlot(human, reduction = "spca_projections", dims = c(1, i), group.by = "Genotype")
ggsave(filename = paste0(home_dir, "/pc_plot_spca/pc_", i, ".pdf"))
}
```

```{r}
human@meta.data
```


```{r}
dittoDimPlot(
  object = human,                   # Your Seurat object
  reduction.use = "spca_projections", # Specify the reduction method
  dim.1 = 3,                        # Specify the first dimension
  dim.2 = 5,                        # Specify the second dimension
  var = "Phase"                  # Metadata variable for coloring
)

dittoPlot(
  object = human,             # Your Seurat object
  var = "G2M.Score",          # The y-axis variable
  group.by = "Genotype"       # Specify that you want a violin plot
)
```

```{r}
wilcox.test(G2M.Score ~ Genotype, data = human@meta.data)
```

```{r}
cor.test(
  ~ G2M.Score + spcaproj_5,  
  data = human@meta.data,          
  method = "spearman"        
)
```




```{r}
p <- DimPlot(human, reduction = "spca_projections", dims = c(3, 5), group.by = "Genotype")
cairo_ps(filename= paste0(home_dir, "/Manuscript and Figures/spc5_plot.eps"), width = 8, height = 7) # name the file as an EPS
p #variable of the figure that is defined/saved, can be used any name
dev.off()
```


### Deviances
```{r}
final_model <- readRDS(file = paste0(home_dir, "/AD_human_objects_ortholog/pca_glm_aic_optimized.RData"))

# Deviance barplot visualization
anova_df <- anova(final_model)
anova_df <- anova_df %>%
  arrange(desc(Deviance))
anova_df <- anova_df %>%
  filter(!row.names(anova_df) %in% c("NULL"))
anova_df

p2 <- ggplot(anova_df, aes(x = reorder(row.names(anova_df), Deviance), y = Deviance)) + 
  geom_bar(stat = "identity") + coord_flip() +
  scale_y_log10() + 
  labs(x = "Principal Component",
       title = "Significant PCA Loadings") +
   theme(axis.text.y = element_text(size = 10))

ggsave(filename= paste0(home_dir, "/Manuscript and Figures/Deviance_PCA.png"), width = 9, height = 7) # name the file as an EPS
p2 #variable of the figure that is defined/saved, can be used any name
dev.off()
```



```{r}
sessionInfo()
```

