---
title: "PC Variance Explained"
author: "Alexander Bergendorf"
date: "2023-10-30"
output: html_document
---

# Library Import
```{r}
library(Seurat)
library(mixOmics)
library(tidyverse)
library(orthogene)
library(dittoSeq)
library(openxlsx)
library(ggrepel)
home_dir <- "C:/Users/aberg/OneDrive/Desktop/Research_Brubaker/AD TransComp-R Project Summer-Fall 2023/Brubaker Summer Research 2023"
#home_dir <- "C:/Users/Alexander/OneDrive/Desktop/Research_Brubaker/AD TransComp-R Project Summer-Fall 2023/Brubaker Summer Research 2023"
```

# sPCA
## All PCs
```{r}
# Loading mouse object with spca loadings
mouse_spca <- readRDS(file = paste0(home_dir, "/AD_mouse_objects_ortholog/mouse_spca_loadings.RData"))

# Loading human object with spca projections
human <- readRDS(file = paste0(home_dir, "/AD_human_objects_ortholog/human_mouse_proj_spca.RData"))
```


```{r}
# Getting mouse loadings
mouse_loading_f <- data.frame(mouse_spca@reductions$spca@feature.loadings)

# Getting human gene expression matrix
t_filt_human_ex <- data.frame(human@assays$RNA@scale.data)
```

```{r}
# Filtering human gene expression matrix to only include the 3000 genes in mouse matrix
ortholog_mat <- convert_orthologs(gene_df = row.names(t_filt_human_ex),
                                    input_species = "human",
                                    output_species = "mouse",
                                    non121_strategy = "drop_both_species")

ortholog_mat <- ortholog_mat %>%
  filter(row.names(ortholog_mat) %in% row.names(mouse_loading_f))
t_filt_human_ex <- t_filt_human_ex %>%
  filter(row.names(t_filt_human_ex) %in% ortholog_mat$input_gene)
row.names(t_filt_human_ex) <- row.names(ortholog_mat)
mouse_loading_f  <- mouse_loading_f %>% arrange(row.names(mouse_loading_f))
t_filt_human_ex <- t_filt_human_ex %>%
  arrange(row.names(t_filt_human_ex))
```


```{r}
t_filt_human_ex <- t(t_filt_human_ex)
t_filt_human_ex <- as.matrix(t_filt_human_ex)
mouse_loading_f <- as.matrix(mouse_loading_f)
ncol(mouse_loading_f)
```


```{r}
# Get the variance of the spca projections
var_mh <- lapply(1:ncol(mouse_loading_f), function(x) (t(mouse_loading_f[,x])%*%t(t_filt_human_ex)%*%t_filt_human_ex%*%mouse_loading_f[,x])/(sum(diag(t(mouse_loading_f)%*%t(t_filt_human_ex)%*%t_filt_human_ex%*%mouse_loading_f))))
```

```{r}
variance_vec <- data.frame(variance = unlist(var_mh))
names_vec <- colnames(mouse_spca@reductions$spca@feature.loadings)
variance_vec$principal_component <- names_vec
write.csv(variance_vec, file = paste0(home_dir, "/spca_variances.csv"))
variance_vec <- read.csv(file = paste0(home_dir, "/spca_variances.csv"))
variance_vec$cum_var <- cumsum(variance_vec$variance)
variance_vec <- variance_vec %>%
  mutate(observation = row_number())
variance_vec
```

```{r}
variance_vec_human <- read.csv(file = paste0(home_dir, "/spca_variances.csv"), row.names = 1)
variance_vec_human$cum_var <- cumsum(variance_vec_human$variance)
variance_vec_human$observation <- 1:nrow(variance_vec_human) 
variance_vec_human

variance_vec_mouse <- read.csv(paste0(home_dir, "/spca_variances_mouse.csv"))
colnames(variance_vec_mouse) <- c("sPC", "sdev")
variance_vec_mouse$cum_var <- cumsum(variance_vec_mouse$sdev)
variance_vec_mouse$variance <- variance_vec_mouse$sdev / sum(variance_vec_mouse$sdev)
variance_vec_mouse$observation <- 1:nrow(variance_vec_mouse) 
variance_vec_mouse


new_df <- data.frame(principal_components = variance_vec_human$principal_component,
                     human_var = variance_vec_human$variance,
                     mouse_var = variance_vec_mouse$variance)


new_df <- new_df[order(-new_df$human_var), ]
top_5 <- head(new_df, 5)
new_df
write.csv(new_df, file = paste0(home_dir, "/mouse_human_var_spca.csv"))
```

```{r}
p <- ggplot(new_df, aes(x = 100 * mouse_var, y = 100 * human_var)) +  # Multiply both variables by 100
  geom_point(size = 3) +  # Increase dot size
  geom_abline(intercept = 0, slope = 1, linetype = "dashed", color = "black") +  # Add diagonal dashed line
  scale_x_continuous(limits = c(0, 100 * max(new_df$mouse_var, new_df$human_var)), name = "Var Exp in Mouse (%)") +
  scale_y_continuous(limits = c(0, 100 * max(new_df$mouse_var, new_df$human_var)), name = "Var Exp in Human (%)") +
  theme_classic() + 
  theme_bw() +
  annotate("text", x = 100 * top_5$mouse_var, y = 100 * top_5$human_var, label = top_5$principal_components, hjust = -0.4, vjust = -0.4, size = 4)

cairo_ps(filename= paste0(home_dir, "/Manuscript and Figures/Figure 4/4B.png"), width = 6, height = 6) # name the file as an EPS
p #variable of the figure that is defined/saved, can be used any name
dev.off()
```

```{r}
p
```


## All PCs encapsulating 80% variance
```{r}
# Getting mouse loadings
mouse_loading_f <- data.frame(mouse_spca@reductions$spca@feature.loadings)

# Getting human gene expression matrix
t_filt_human_ex <- data.frame(human@assays$RNA@scale.data)
```

```{r}
# Filtering human gene expression matrix to only include the genes in mouse matrix
ortholog_mat <- convert_orthologs(gene_df = row.names(t_filt_human_ex),
                                    input_species = "human",
                                    output_species = "mouse",
                                    non121_strategy = "drop_both_species")

ortholog_mat <- ortholog_mat %>%
  filter(row.names(ortholog_mat) %in% row.names(mouse_loading_f))
t_filt_human_ex <- t_filt_human_ex %>%
  filter(row.names(t_filt_human_ex) %in% ortholog_mat$input_gene)
row.names(t_filt_human_ex) <- row.names(ortholog_mat)
mouse_loading_f  <- mouse_loading_f %>% arrange(row.names(mouse_loading_f))
t_filt_human_ex <- t_filt_human_ex %>%
  arrange(row.names(t_filt_human_ex))
```

```{r}
mouse_loading_f <- mouse_loading_f[,1:37]
t_filt_human_ex <- t(t_filt_human_ex)
t_filt_human_ex <- as.matrix(t_filt_human_ex)
mouse_loading_f <- as.matrix(mouse_loading_f)
```


```{r}
# Get the variance of the spca projections
var_mh <- lapply(1:ncol(mouse_loading_f), function(x) (t(mouse_loading_f[,x])%*%t(t_filt_human_ex)%*%t_filt_human_ex%*%mouse_loading_f[,x])/(sum(diag(t(mouse_loading_f)%*%t(t_filt_human_ex)%*%t_filt_human_ex%*%mouse_loading_f))))
```


```{r}
variance_vec <- data.frame(variance = unlist(var_mh))
names_vec <- colnames(mouse_spca@reductions$spca@feature.loadings)[1:37]
variance_vec$principal_component <- names_vec
write.csv(variance_vec, file = paste0(home_dir, "/spca_variances_80_var.csv"))
variance_vec <- read.csv(file = paste0(home_dir, "/spca_variances_80_var.csv"), row.names = 1)
variance_vec$cum_var <- cumsum(variance_vec$variance)
variance_vec <- variance_vec %>%
  mutate(observation = row_number())
variance_vec
```






```{r}
# Variance explained of projections
ggplot(data = variance_vec, aes(x = observation, y = cum_var, group = 1)) +
  geom_line(color = "black") +  # Changed line color to blue
  geom_point(color = "black") +  # Added points with the same color
  geom_hline(yintercept = 0.80, linetype = "dashed", color = "red") +  # Added red dashed line at 0.80 cum_var
  geom_point(data = subset(variance_vec, cum_var == 0.80), aes(x = observation, y = cum_var), color = "red", size = 3) +  # Add red point at the threshold
  ylab("Cumulative Variance") +
  xlab("sPC") +# Changed y-axis label
  theme_classic() + theme_bw()

```

```{r}
variance_vec_human <- read.csv(file = paste0(home_dir, "/spca_variances_80_var.csv"), row.names = 1)
variance_vec_human$cum_var <- cumsum(variance_vec_human$variance)
variance_vec_human$observation <- 1:nrow(variance_vec_human) 
variance_vec_human

variance_vec_mouse <- read.csv(paste0(home_dir, "/spca_variances_mouse.csv"))
colnames(variance_vec_mouse) <- c("sPC", "sdev")
variance_vec_mouse$cum_var <- cumsum(variance_vec_mouse$sdev)
variance_vec_mouse <- variance_vec_mouse[1:37,]
variance_vec_mouse$variance <- variance_vec_mouse$sdev / variance_vec_mouse$cum_var[37]
variance_vec_mouse$observation <- 1:nrow(variance_vec_mouse) 
variance_vec_mouse


new_df <- data.frame(principal_components = variance_vec_human$principal_component,
                     human_var = variance_vec_human$variance,
                     mouse_var = variance_vec_mouse$variance)


new_df <- new_df[order(-new_df$human_var), ]
top_5 <- head(new_df, 5)
write.csv(new_df, file = paste0(home_dir, "/mouse_human_var_80_spca.csv"))
sum(new_df$mouse_var)
new_df
```

```{r}
ggplot(new_df, aes(x = 100 * mouse_var, y = 100 * human_var)) +  # Multiply both variables by 100
  geom_point(size = 3) +  # Increase dot size
  geom_abline(intercept = 0, slope = 1, linetype = "dashed", color = "black") +  # Add diagonal dashed line
  scale_x_continuous(limits = c(0, 100 * max(new_df$mouse_var, new_df$human_var)), name = "Var Exp in Mouse (%)") +
  scale_y_continuous(limits = c(0, 100 * max(new_df$mouse_var, new_df$human_var)), name = "Var Exp in Human (%)") +
  theme_classic() + 
  theme_bw() +
  annotate("text", x = 100 * top_5$mouse_var, y = 100 * top_5$human_var, label = top_5$principal_components, hjust = -0.1, vjust = -0.2, size = 3)
ggsave('C:/Users/aberg/OneDrive/Desktop/human_mouse_variance_spca_80.png', width = 5, height = 4)
```


## Only with the AIC optimized PCs
```{r}
# Loading mouse object with spca loadings
mouse_spca <- readRDS(file = paste0(home_dir, "/AD_mouse_objects_ortholog/mouse_spca_loadings.RData"))

# Loading human object with spca projections
human <- readRDS(file = paste0(home_dir, "/AD_human_objects_ortholog/human_mouse_proj_spca.RData"))
```

```{r}
final_model <- readRDS(file = paste0(home_dir, "/AD_human_objects_ortholog/spca_glm_aic_optimized.RData"))
summary(final_model)
```


```{r}
# Getting mouse loadings
mouse_loading_f <- data.frame(mouse_spca@reductions$spca@feature.loadings)

# Getting human gene expression matrix
t_filt_human_ex <- data.frame(human@assays$RNA@scale.data)
```

```{r}
# Filtering human gene expression matrix to only include the 3000 genes in mouse matrix
ortholog_mat <- convert_orthologs(gene_df = row.names(t_filt_human_ex),
                                    input_species = "human",
                                    output_species = "mouse",
                                    non121_strategy = "drop_both_species")

ortholog_mat <- ortholog_mat %>%
  filter(row.names(ortholog_mat) %in% row.names(mouse_loading_f))
t_filt_human_ex <- t_filt_human_ex %>%
  filter(row.names(t_filt_human_ex) %in% ortholog_mat$input_gene)
row.names(t_filt_human_ex) <- row.names(ortholog_mat)
mouse_loading_f  <- mouse_loading_f %>% arrange(row.names(mouse_loading_f))
t_filt_human_ex <- t_filt_human_ex %>%
  arrange(row.names(t_filt_human_ex))
```


```{r}
t_filt_human_ex <- t(t_filt_human_ex)
t_filt_human_ex <- as.matrix(t_filt_human_ex)
mouse_loading_f <- mouse_loading_f[,c(2, 3, 4, 5, 6, 7, 9, 10, 11, 12, 13,
                   14, 15, 16, 18, 20, 21, 23, 24, 25, 27, 28, 29, 31, 33, 37)]
mouse_loading_f <- as.matrix(mouse_loading_f)
ncol(mouse_loading_f)
```


```{r}
# Get the variance of the spca projections
var_mh <- lapply(1:ncol(mouse_loading_f), function(x) (t(mouse_loading_f[,x])%*%t(t_filt_human_ex)%*%t_filt_human_ex%*%mouse_loading_f[,x])/(sum(diag(t(mouse_loading_f)%*%t(t_filt_human_ex)%*%t_filt_human_ex%*%mouse_loading_f))))
```

```{r}
variance_vec <- data.frame(variance = unlist(var_mh))
names_vec <- colnames(mouse_spca@reductions$spca@feature.loadings)[c(2, 3, 4, 5, 6, 7, 9, 10, 11, 12, 13,
                   14, 15, 16, 18, 20, 21, 23, 24, 25, 27, 28, 29, 31, 33, 37)]
variance_vec$principal_component <- names_vec
write.csv(variance_vec, file = paste0(home_dir, "/spca_variances_aic_optim.csv"))
variance_vec$cum_var <- cumsum(variance_vec$variance)
variance_vec <- variance_vec %>%
  mutate(observation = row_number())
variance_vec
```

```{r}
variance_vec_human <- read.csv(file = paste0(home_dir, "/spca_variances_aic_optim.csv"), row.names = 1)
variance_vec_human$cum_var <- cumsum(variance_vec_human$variance)
variance_vec_human$observation <- 1:nrow(variance_vec_human) 
variance_vec_human

variance_vec_mouse <- read.csv(paste0(home_dir, "/spca_variances_mouse.csv"))
colnames(variance_vec_mouse) <- c("sPC", "sdev")
variance_vec_mouse$cum_var <- cumsum(variance_vec_mouse$sdev)
variance_vec_mouse <- variance_vec_mouse[c(2, 3, 4, 5, 6, 7, 9, 10, 11, 12, 13,
                   14, 15, 16, 18, 20, 21, 23, 24, 25, 27, 28, 29, 31, 33, 37),]
variance_vec_mouse$variance <- variance_vec_mouse$sdev / sum(variance_vec_mouse$sdev)
variance_vec_mouse$observation <- 1:nrow(variance_vec_mouse) 
variance_vec_mouse


new_df <- data.frame(principal_components = variance_vec_human$principal_component,
                     human_var = variance_vec_human$variance,
                     mouse_var = variance_vec_mouse$variance)


new_df <- new_df[order(-new_df$human_var), ]
top_5 <- head(new_df, 5)
new_df
write.csv(new_df, file = paste0(home_dir, "/mouse_human_var_aic_optim_spca.csv"))
```

```{r}
p <- ggplot(new_df, aes(x = 100 * mouse_var, y = 100 * human_var)) +  
  geom_point(size = 3) +  
  geom_abline(intercept = 0, slope = 1, linetype = "dashed", color = "black") +  
  scale_x_continuous(limits = c(0, 100 * max(new_df$mouse_var, new_df$human_var)), name = "Var Exp in Mouse (%)") +
  scale_y_continuous(limits = c(0, 100 * max(new_df$mouse_var, new_df$human_var)), name = "Var Exp in Human (%)") +
  theme_classic() + 
  theme_bw() +
  geom_text_repel(data = top_5, aes(x = 100 * mouse_var, y = 100 * human_var, label = principal_components), 
                  hjust = -0.4, vjust = -0.4, size = 4, box.padding = 0.35, point.padding = 0.3, max.overlaps = 10)

# Save the plot as PNG
png(filename= paste0(home_dir, "/Manuscript and Figures/Figure 4/4B.png"), width = 6, height = 6, units = "in", res = 300) 
p  # Plot the figure
dev.off()
```


# PCA
## All PCs
```{r}
# Loading mouse object with pca loadings
mouse_pca <- readRDS(file = paste0(home_dir, "/AD_mouse_objects_ortholog/mouse_pca_loadings.RData"))

# loading human object with pca projections
human <- readRDS(file = paste0(home_dir, "/AD_human_objects_ortholog/human_mouse_proj_pca.RData"))
```


```{r}
# Getting mouse loadings
mouse_loading_f <- data.frame(mouse_pca@reductions$pca@feature.loadings)

# Getting human gene expression matrix
t_filt_human_ex <- data.frame(human@assays$RNA@scale.data)
```

```{r}
# Filtering human gene expression matrix to only include the 3000 genes in mouse matrix
ortholog_mat <- convert_orthologs(gene_df = row.names(t_filt_human_ex),
                                    input_species = "human",
                                    output_species = "mouse",
                                    non121_strategy = "drop_both_species")

ortholog_mat <- ortholog_mat %>%
  filter(row.names(ortholog_mat) %in% row.names(mouse_loading_f))
t_filt_human_ex <- t_filt_human_ex %>%
  filter(row.names(t_filt_human_ex) %in% ortholog_mat$input_gene)
row.names(t_filt_human_ex) <- row.names(ortholog_mat)
mouse_loading_f  <- mouse_loading_f %>% arrange(row.names(mouse_loading_f))
t_filt_human_ex <- t_filt_human_ex %>%
  arrange(row.names(t_filt_human_ex))
```

```{r}
t_filt_human_ex <- t(t_filt_human_ex)
t_filt_human_ex <- as.matrix(t_filt_human_ex)
mouse_loading_f <- as.matrix(mouse_loading_f)
ncol(mouse_loading_f)
```


```{r}
# Calculate variances of human pca
var_mh <- lapply(1:ncol(mouse_loading_f), function(x) (t(mouse_loading_f[,x])%*%t(t_filt_human_ex)%*%t_filt_human_ex%*%mouse_loading_f[,x])/(sum(diag(t(mouse_loading_f)%*%t(t_filt_human_ex)%*%t_filt_human_ex%*%mouse_loading_f))))
```

```{r}
variance_vec <- data.frame(variance = unlist(var_mh))
names_vec <- colnames(mouse_pca@reductions$pca@feature.loadings)
variance_vec$principal_component <- names_vec
write.csv(variance_vec, file = paste0(home_dir, "/pca_variances.csv"))
variance_vec$cum_var <- cumsum(variance_vec$variance)
variance_vec <- variance_vec %>%
  mutate(observation = row_number())
variance_vec
```

```{r}
variance_vec_human <- read.csv(file = paste0(home_dir, "/pca_variances.csv"), row.names = 1)
variance_vec_human$cum_var <- cumsum(variance_vec_human$variance)
variance_vec_human$observation <- 1:nrow(variance_vec_human) 
variance_vec_human

variance_vec_mouse <- read.csv(paste0(home_dir, "/pca_variances_mouse.csv"))
colnames(variance_vec_mouse) <- c("PC", "sdev")
variance_vec_mouse$cum_var <- cumsum(variance_vec_mouse$sdev)
variance_vec_mouse$variance <- variance_vec_mouse$sdev / sum(variance_vec_mouse$sdev)
variance_vec_mouse$observation <- 1:nrow(variance_vec_mouse) 
variance_vec_mouse


new_df <- data.frame(principal_components = variance_vec_human$principal_component,
                     human_var = variance_vec_human$variance,
                     mouse_var = variance_vec_mouse$variance)


new_df <- new_df[order(-new_df$human_var), ]
top_5 <- head(new_df, 5)
new_df
write.csv(new_df, file = paste0(home_dir, "/mouse_human_var_pca.csv"))
```

```{r}
p <- ggplot(new_df, aes(x = 100 * mouse_var, y = 100 * human_var)) +  
  geom_point(size = 3) +  
  geom_abline(intercept = 0, slope = 1, linetype = "dashed", color = "black") +  
  scale_x_continuous(limits = c(0, 100 * max(new_df$mouse_var, new_df$human_var)), name = "Var Exp in Mouse (%)") +
  scale_y_continuous(limits = c(0, 100 * max(new_df$mouse_var, new_df$human_var)), name = "Var Exp in Human (%)") +
  theme_classic() + 
  theme_bw() +
  annotate("text", x = 100 * top_5$mouse_var, y = 100 * top_5$human_var, label = top_5$principal_components, 
           hjust = -0.4, vjust = -0.4, size = 4)

# Save the plot as PNG
png(filename= paste0(home_dir, "/Manuscript and Figures/Figure 4/4A.png"), width = 6, height = 6, units = "in", res = 300) 
p  # Plot the figure
dev.off() 
```

```{r}
p
```


## All PCs encapsulating 80% variance
```{r}
# Getting mouse loadings
mouse_loading_f <- data.frame(mouse_pca@reductions$pca@feature.loadings)

# Getting human gene expression matrix
t_filt_human_ex <- data.frame(human@assays$RNA@scale.data)
```


```{r}
# Filtering human gene expression matrix to only include the 3000 genes in mouse matrix
ortholog_mat <- convert_orthologs(gene_df = row.names(t_filt_human_ex),
                                    input_species = "human",
                                    output_species = "mouse",
                                    non121_strategy = "drop_both_species")

ortholog_mat <- ortholog_mat %>%
  filter(row.names(ortholog_mat) %in% row.names(mouse_loading_f))
t_filt_human_ex <- t_filt_human_ex %>%
  filter(row.names(t_filt_human_ex) %in% ortholog_mat$input_gene)
row.names(t_filt_human_ex) <- row.names(ortholog_mat)
mouse_loading_f  <- mouse_loading_f %>% arrange(row.names(mouse_loading_f))
t_filt_human_ex <- t_filt_human_ex %>%
  arrange(row.names(t_filt_human_ex))
```



```{r}
mouse_loading_f <- mouse_loading_f[,1:36]
t_filt_human_ex <- t(t_filt_human_ex)
t_filt_human_ex <- as.matrix(t_filt_human_ex)
mouse_loading_f <- as.matrix(mouse_loading_f)
```


```{r}
# Calculate variances of human pca
var_mh <- lapply(1:ncol(mouse_loading_f), function(x) (t(mouse_loading_f[,x])%*%t(t_filt_human_ex)%*%t_filt_human_ex%*%mouse_loading_f[,x])/(sum(diag(t(mouse_loading_f)%*%t(t_filt_human_ex)%*%t_filt_human_ex%*%mouse_loading_f))))
```

```{r}
variance_vec <- data.frame(variance = unlist(var_mh))
names_vec <- colnames(mouse_pca@reductions$pca@feature.loadings)[1:36]
variance_vec$principal_component <- names_vec
write.csv(variance_vec, file = paste0(home_dir, "/pca_variances_80_var.csv"))
variance_vec$cum_var <- cumsum(variance_vec$variance)
variance_vec <- variance_vec %>%
  mutate(observation = row_number())
variance_vec
```


```{r}
ggsave('C:/Users/aberg/OneDrive/Desktop/human_variance_explained_pca_all.png', width = 5, height = 4)
# Variance explained of projections
ggplot(data = variance_vec, aes(x = observation, y = cum_var, group = 1)) +
  geom_line(color = "black") +  # Changed line color to blue
  geom_point(color = "black") +  # Added points with the same color
  geom_hline(yintercept = 0.80, linetype = "dashed", color = "red") +  # Added red dashed line at 0.80 cum_var
  geom_point(data = subset(variance_vec, cum_var == 0.80), aes(x = observation, y = cum_var), color = "red", size = 3) +  # Add red point at the threshold
  ylab("Cumulative Variance") +# Changed y-axis label
  xlab("PC") +
  theme_classic() + theme_bw()
dev.off()
```

```{r}
variance_vec_human <- read.csv(file = paste0(home_dir, "/pca_variances_80_var.csv"), row.names = 1)
variance_vec_human$cum_var <- cumsum(variance_vec_human$variance)
variance_vec_human$observation <- 1:nrow(variance_vec_human) 
variance_vec_human

variance_vec_mouse <- read.csv(paste0(home_dir, "/pca_variances_mouse.csv"))
colnames(variance_vec_mouse) <- c("PC", "sdev")
variance_vec_mouse$cum_var <- cumsum(variance_vec_mouse$sdev)
variance_vec_mouse <- variance_vec_mouse[1:36,]
variance_vec_mouse$variance <- variance_vec_mouse$sdev / variance_vec_mouse$cum_var[36]
variance_vec_mouse$observation <- 1:nrow(variance_vec_mouse) 
variance_vec_mouse


new_df <- data.frame(principal_components = variance_vec_human$principal_component,
                     human_var = variance_vec_human$variance,
                     mouse_var = variance_vec_mouse$variance)


new_df <- new_df[order(-new_df$human_var), ]
top_5 <- head(new_df, 5)
write.csv(new_df, file = paste0(home_dir, "/mouse_human_var_80.csv"))
```

```{r}
ggplot(new_df, aes(x = 100 * mouse_var, y = 100 * human_var)) +  # Multiply both variables by 100
  geom_point(size = 3) +  # Increase dot size
  geom_abline(intercept = 0, slope = 1, linetype = "dashed", color = "black") +  # Add diagonal dashed line
  scale_x_continuous(limits = c(0, 100 * max(new_df$mouse_var, new_df$human_var)), name = "Var Exp in Mouse (%)") +
  scale_y_continuous(limits = c(0, 100 * max(new_df$mouse_var, new_df$human_var)), name = "Var Exp in Human (%)") +
  theme_classic() + 
  theme_bw() +
  annotate("text", x = 100 * top_5$mouse_var, y = 100 * top_5$human_var, label = top_5$principal_components, hjust = -0.1, vjust = -0.2, size = 3)
ggsave('C:/Users/aberg/OneDrive/Desktop/human_mouse_variance_pca_80.png', width = 5, height = 4)
```

## Only with the AIC optimized PCs
```{r}
# Loading mouse object with pca loadings
mouse_pca <- readRDS(file = paste0(home_dir, "/AD_mouse_objects_ortholog/mouse_pca_loadings.RData"))

# loading human object with pca projections
human <- readRDS(file = paste0(home_dir, "/AD_human_objects_ortholog/human_mouse_proj_pca.RData"))
```

```{r}
final_model <- readRDS(file = paste0(home_dir, "/AD_human_objects_ortholog/pca_glm_aic_optimized.RData"))
summary(final_model)
```


```{r}
# Getting mouse loadings
mouse_loading_f <- data.frame(mouse_pca@reductions$pca@feature.loadings)

# Getting human gene expression matrix
t_filt_human_ex <- data.frame(human@assays$RNA@scale.data)
```

```{r}
# Filtering human gene expression matrix to only include the 3000 genes in mouse matrix
ortholog_mat <- convert_orthologs(gene_df = row.names(t_filt_human_ex),
                                    input_species = "human",
                                    output_species = "mouse",
                                    non121_strategy = "drop_both_species")

ortholog_mat <- ortholog_mat %>%
  filter(row.names(ortholog_mat) %in% row.names(mouse_loading_f))
t_filt_human_ex <- t_filt_human_ex %>%
  filter(row.names(t_filt_human_ex) %in% ortholog_mat$input_gene)
row.names(t_filt_human_ex) <- row.names(ortholog_mat)
mouse_loading_f  <- mouse_loading_f %>% arrange(row.names(mouse_loading_f))
t_filt_human_ex <- t_filt_human_ex %>%
  arrange(row.names(t_filt_human_ex))
```

```{r}
t_filt_human_ex <- t(t_filt_human_ex)
t_filt_human_ex <- as.matrix(t_filt_human_ex)
mouse_loading_f <- mouse_loading_f[,c(1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13,
                   15, 18, 21, 22, 27, 29, 30, 32, 36)]
mouse_loading_f <- as.matrix(mouse_loading_f)
ncol(mouse_loading_f)
```


```{r}
# Calculate variances of human pca
var_mh <- lapply(1:ncol(mouse_loading_f), function(x) (t(mouse_loading_f[,x])%*%t(t_filt_human_ex)%*%t_filt_human_ex%*%mouse_loading_f[,x])/(sum(diag(t(mouse_loading_f)%*%t(t_filt_human_ex)%*%t_filt_human_ex%*%mouse_loading_f))))
```

```{r}
variance_vec <- data.frame(variance = unlist(var_mh))
names_vec <- colnames(mouse_pca@reductions$pca@feature.loadings)[c(1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13,
                   15, 18, 21, 22, 27, 29, 30, 32, 36)]
variance_vec$principal_component <- names_vec
write.csv(variance_vec, file = paste0(home_dir, "/pca_variances.csv"))
variance_vec$cum_var <- cumsum(variance_vec$variance)
variance_vec <- variance_vec %>%
  mutate(observation = row_number())
variance_vec
```

```{r}
variance_vec_human <- read.csv(file = paste0(home_dir, "/pca_variances.csv"), row.names = 1)
variance_vec_human$cum_var <- cumsum(variance_vec_human$variance)
variance_vec_human$observation <- 1:nrow(variance_vec_human) 
variance_vec_human

variance_vec_mouse <- read.csv(paste0(home_dir, "/pca_variances_mouse.csv"))
colnames(variance_vec_mouse) <- c("PC", "sdev")
variance_vec_mouse$cum_var <- cumsum(variance_vec_mouse$sdev)
variance_vec_mouse <- variance_vec_mouse[c(1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13,
                   15, 18, 21, 22, 27, 29, 30, 32, 36),]
variance_vec_mouse$variance <- variance_vec_mouse$sdev / sum(variance_vec_mouse$sdev)
variance_vec_mouse$observation <- 1:nrow(variance_vec_mouse) 
variance_vec_mouse


new_df <- data.frame(principal_components = variance_vec_human$principal_component,
                     human_var = variance_vec_human$variance,
                     mouse_var = variance_vec_mouse$variance)


new_df <- new_df[order(-new_df$human_var), ]
top_5 <- head(new_df, 5)
new_df
write.csv(new_df, file = paste0(home_dir, "/mouse_human_var_pca.csv"))
```

```{r}
p <- ggplot(new_df, aes(x = 100 * mouse_var, y = 100 * human_var)) +  # Multiply both variables by 100
  geom_point(size = 3) +  # Increase dot size
  geom_abline(intercept = 0, slope = 1, linetype = "dashed", color = "black") +  # Add diagonal dashed line
  scale_x_continuous(limits = c(0, 100 * max(new_df$mouse_var, new_df$human_var)), name = "Var Exp in Mouse (%)") +
  scale_y_continuous(limits = c(0, 100 * max(new_df$mouse_var, new_df$human_var)), name = "Var Exp in Human (%)") +
  theme_classic() + 
  theme_bw() +
  annotate("text", x = 100 * top_5$mouse_var, y = 100 * top_5$human_var, label = top_5$principal_components, hjust = -0.4, vjust = -0.4, size = 2.5)

cairo_ps(filename= paste0(home_dir, "/Manuscript and Figures/Figure 4/4A.eps"), width = 6, height = 6) # name the file as an EPS
p #variable of the figure that is defined/saved, can be used any name
dev.off()
```


# Session Info
```{r}
sessionInfo()
```

